modules/internal/ChapelDomain.chpl

```diff
@@ -1283,6 +1283,271 @@ module ChapelDomain {
+
+      /*
+        Check if a given class reference is ``nil`` without triggering
+        runtime nilability checks.
+      */
+      proc isClassReferenceNil(const ref x) {
+        if isClassType(x.type) {
+          return if isBorrowedOrUnmanagedClassType(x.type)
+            then x == nil
+            else x.chpl_p == nil;
+        } else {
+          return false;
+        }
+      }
+
+      /*
+        Return ``true`` if the value at a given index in an array has
+        been initialized.
+      */
+      proc isElementInitialized(arr: [?d], idx) {
+        if _checks {
+          type T = arr.eltType;
+          if isNonNilableClassType(T) {
+            var actualArr = chpl__getActualArray(arr);
+            return !actualArr.chpl_unsafeAssignIsClassElementNil(this, idx);
+          } else if !isDefaultInitializable(T) {
+            halt('Not implemented yet!');
+            return false;
+          } else {
+            return true;
+          }
+
+        // No checks, so just issue a compiler error.
+        } else {
+          param msg = 'Cannot check initialization state of non-' +
+                      'default-initializable array elements from a ' +
+                      'manager initialized with \'checks=false\'';
+          compilerError(msg);
+          return false;
+        }
+      }
+
+      pragma "no doc"
+      proc _checkIfAllElementsAreInitialized(baseArr) {
+
+        // Arrays of non-nilable classes.
+        if baseArr.chpl_isElementTypeNonNilableClass() {
+          for idx in newIndices() {
+            if baseArr.chpl_unsafeAssignIsClassElementNil(this, idx) {
+              baseArr.chpl_unsafeAssignHaltUninitializedElement(idx);
+              halt();
+            }
+          }
+
+        // Arrays of non-default-initializable types.
+        } else if !baseArr.chpl_isElementTypeDefaultInitializable() {
+
+          // TODO: For non-default-initializable elements, we'd iterate
+          // through a bitvector mapped to newly added slots.
+          halt('Not implemented yet!');
+
+        // Should never reach here.
+        } else {
+          halt('Internal error!');
+        }
+      }
+
+      pragma "no doc"
+      proc _ensureNoLongerManagingThis() {
+        if !_isActiveManager then return; else _isActiveManager = false;
+
+        // Possible runtime checks, reset the resize policy of owned arrays.
+        for baseArr in _lhsBaseArrays() {
+          if !baseArr.chpl_isElementTypeDefaultInitializable() {
+            if _checks then _checkIfAllElementsAreInitialized(baseArr);
+            const policy = chpl_DdataResizePolicy.Normal;
+            baseArr.chpl_setResizePolicy(policy);
+          }
+        }
+      }
+
+      proc deinit() {
+        _ensureNoLongerManagingThis();
+      }
+
+      pragma "no doc"
+      proc _isArrayOwnedByLhsDomain(arr) {
+        for baseArr in _lhsBaseArrays() do
+          if baseArr == chpl__getActualArray(arr) then return true;
```

Comment:
This should probably be just `return arr.dsiGetBaseDom() == _lhsInstance`

Reply:
Awesome!